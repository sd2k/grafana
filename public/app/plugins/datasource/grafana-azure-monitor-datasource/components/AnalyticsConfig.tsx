import React, { FunctionComponent, useEffect, useMemo, useReducer, useState } from 'react';
import { SelectableValue } from '@grafana/data';
import { AzureCredentialsForm } from './AzureCredentialsForm';
import { InlineFormLabel, LegacyForms, Button } from '@grafana/ui';
const { Select, Switch } = LegacyForms;
import { AzureDataSourceSettings, AzureCredentials } from '../types';
import {
  getCredentials,
  getLogAnalyticsCredentials,
  isCredentialsComplete,
  updateLogAnalyticsCredentials,
  updateLogAnalyticsSameAs,
} from '../credentials';

export interface Props {
  options: AzureDataSourceSettings;
  updateOptions: (optionsFunc: (options: AzureDataSourceSettings) => AzureDataSourceSettings) => void;
  getSubscriptions: () => Promise<Array<SelectableValue<string>>>;
  getWorkspaces: (subscriptionId: string) => Promise<Array<SelectableValue<string>>>;
}

export const AnalyticsConfig: FunctionComponent<Props> = (props: Props) => {
  const { updateOptions, getSubscriptions, getWorkspaces } = props;
  const primaryCredentials = useMemo(() => getCredentials(props.options), [props.options]);
  const logAnalyticsCredentials = useMemo(() => getLogAnalyticsCredentials(props.options), [props.options]);
  const subscriptionId = logAnalyticsCredentials
    ? props.options.jsonData.logAnalyticsSubscriptionId
    : props.options.jsonData.subscriptionId;

  const hasRequiredFields =
    subscriptionId &&
    (logAnalyticsCredentials
      ? isCredentialsComplete(logAnalyticsCredentials)
      : isCredentialsComplete(primaryCredentials));

  const defaultWorkspace = props.options.jsonData.logAnalyticsDefaultWorkspace;

  const [workspaces, setWorkspaces] = useState<SelectableValue[]>([]);
  const [loadWorkspaces, onLoadWorkspaces] = useReducer((val) => val + 1, 0);
  useEffect(() => {
    if (!hasRequiredFields || !subscriptionId) {
      updateWorkspaces([]);
      return;
    }
    let canceled = false;
    getWorkspaces(subscriptionId).then((result) => {
      if (!canceled) {
        updateWorkspaces(result);
      }
    });
    return () => {
      canceled = true;
    };
    // This effect is intended to be called only once initially and on Load Workspaces click
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [loadWorkspaces, subscriptionId]);

  const updateWorkspaces = (received: Array<SelectableValue<string>>) => {
    setWorkspaces(received);
    if (!defaultWorkspace && received.length > 0) {
      // Setting the default workspace if workspaces received but no default workspace selected
      updateOptions((options) => {
        return {
          ...options,
          jsonData: {
            ...options.jsonData,
            logAnalyticsDefaultWorkspace: received[0].value,
          },
        };
      });
    } else if (defaultWorkspace) {
      const found = received.find((opt) => opt.value === defaultWorkspace);
      if (!found) {
        // Unsetting the default workspace if it isn't found among the received workspaces
        updateOptions((options) => {
          return {
            ...options,
            jsonData: {
              ...options.jsonData,
              logAnalyticsDefaultWorkspace: undefined,
            },
          };
        });
      }
    }
  };

  const [sameAsSwitched, setSameAsSwitched] = useState(false);

  const onCredentialsChange = (updatedCredentials: AzureCredentials) => {
    updateOptions((options) => updateLogAnalyticsCredentials(options, updatedCredentials));
  };

  const onLogAnalyticsSameAsChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const sameAs = event.currentTarget.checked;
    updateOptions((options) => updateLogAnalyticsSameAs(options, sameAs));
    setSameAsSwitched(true);
  };

  const onLogAnalyticsDefaultSubscriptionChange = (subscriptionId: string | undefined) => {
    updateOptions((options) => {
      return {
        ...options,
        jsonData: {
          ...options.jsonData,
          logAnalyticsSubscriptionId: subscriptionId || '',
        },
      };
    });
  };

  const onDefaultWorkspaceChange = (selected: SelectableValue<string>) => {
    updateOptions((options) => {
      return {
        ...options,
        jsonData: {
          ...options.jsonData,
          logAnalyticsDefaultWorkspace: selected.value || '',
        },
      };
    });
  };

  const tooltipAttribute = {
    ...(!logAnalyticsCredentials && {
      tooltip: 'Workspaces are pulled from default subscription selected above.',
    }),
  };

  const showSameAsHelpMsg = sameAsSwitched && !primaryCredentials.clientSecret;

  return (
    <>
      <h3 className="page-heading">Azure Monitor Logs Details</h3>
      <Switch
        label="Same details as Azure Monitor API"
        checked={!logAnalyticsCredentials}
        onChange={onLogAnalyticsSameAsChange}
        {...tooltipAttribute}
      />
      {showSameAsHelpMsg && (
        <div className="grafana-info-box m-t-2">
          <div className="alert-body">
            <p>Re-enter your Azure Monitor Client Secret to use this setting.</p>
          </div>
        </div>
      )}
      {logAnalyticsCredentials && (
        <AzureCredentialsForm
          credentials={logAnalyticsCredentials}
          defaultSubscription={subscriptionId}
          onCredentialsChange={onCredentialsChange}
          onDefaultSubscriptionChange={onLogAnalyticsDefaultSubscriptionChange}
          getSubscriptions={getSubscriptions}
        />
      )}
      <div className="gf-form-group">
        <div className="gf-form-inline">
          <div className="gf-form">
            <InlineFormLabel
              className="width-12"
              tooltip="Choose the default/preferred Workspace for Azure Log Analytics queries."
            >
              Default Workspace
            </InlineFormLabel>
            <div className="width-25">
              <Select
                value={workspaces.find((opt) => opt.value === defaultWorkspace)}
                options={workspaces}
                onChange={onDefaultWorkspaceChange}
              />
            </div>
          </div>
        </div>
        <div className="gf-form-inline">
          <div className="gf-form">
            <div className="max-width-30 gf-form-inline">
              <Button
                variant="secondary"
                size="sm"
                type="button"
                onClick={onLoadWorkspaces}
                disabled={!hasRequiredFields}
              >
                Load Workspaces
              </Button>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default AnalyticsConfig;
