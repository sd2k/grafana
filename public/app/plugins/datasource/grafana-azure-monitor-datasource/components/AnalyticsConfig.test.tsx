import React from 'react';
import { shallow } from 'enzyme';
import AnalyticsConfig, { Props } from './AnalyticsConfig';

const setup = (propsFunc?: (props: Props) => Props) => {
  let props: Props = {
    options: {
      id: 21,
      uid: 'x',
      orgId: 1,
      name: 'Azure Monitor-10-10',
      type: 'grafana-azure-monitor-datasource',
      typeName: 'Azure',
      typeLogoUrl: '',
      access: 'proxy',
      url: '',
      password: '',
      user: '',
      database: '',
      basicAuth: false,
      basicAuthUser: '',
      basicAuthPassword: '',
      withCredentials: false,
      isDefault: false,
      secureJsonFields: {
        logAnalyticsClientSecret: false,
      },
      jsonData: {
        cloudName: '',
        subscriptionId: '',
        azureLogAnalyticsSameAs: false,
        logAnalyticsDefaultWorkspace: '',
        logAnalyticsTenantId: '',
      },
      secureJsonData: {
        logAnalyticsClientSecret: '',
      },
      version: 1,
      readOnly: false,
    },
    updateOptions: jest.fn(),
    getSubscriptions: jest.fn(),
    getWorkspaces: jest.fn(),
  };

  if (propsFunc) {
    props = propsFunc(props);
  }

  return shallow(<AnalyticsConfig {...props} />);
};

describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should disable log analytics credentials form', () => {
    const wrapper = setup((props) => ({
      ...props,
      options: {
        ...props.options,
        jsonData: {
          ...props.options.jsonData,
          azureLogAnalyticsSameAs: true,
        },
      },
    }));
    expect(wrapper).toMatchSnapshot();
  });

  it('should enable azure log analytics load workspaces button', () => {
    const wrapper = setup((props) => ({
      ...props,
      options: {
        ...props.options,
        jsonData: {
          ...props.options.jsonData,
          azureLogAnalyticsSameAs: false,
          logAnalyticsDefaultWorkspace: '',
          logAnalyticsTenantId: 'e7f3f661-a933-4b3f-8176-51c4f982ec48',
          logAnalyticsClientId: '44693801-6ee6-49de-9b2d-9106972f9572',
          logAnalyticsSubscriptionId: 'e3fe4fde-ad5e-4d60-9974-e2f3562ffdf2',
          logAnalyticsClientSecret: 'cddcc020-2c94-460a-a3d0-df3147ffa792',
        },
      },
    }));
    expect(wrapper).toMatchSnapshot();
  });
});
