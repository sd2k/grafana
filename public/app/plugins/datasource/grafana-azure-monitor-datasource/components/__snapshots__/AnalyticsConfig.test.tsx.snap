// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should disable log analytics credentials form 1`] = `
<Fragment>
  <h3
    className="page-heading"
  >
    Azure Monitor Logs Details
  </h3>
  <Switch
    checked={true}
    label="Same details as Azure Monitor API"
    onChange={[Function]}
    tooltip="Workspaces are pulled from default subscription selected above."
  />
  <div
    className="gf-form-group"
  >
    <div
      className="gf-form-inline"
    >
      <div
        className="gf-form"
      >
        <FormLabel
          className="width-12"
          tooltip="Choose the default/preferred Workspace for Azure Log Analytics queries."
        >
          Default Workspace
        </FormLabel>
        <div
          className="width-25"
        >
          <Select
            allowCustomValue={false}
            autoFocus={false}
            backspaceRemovesValue={true}
            className=""
            components={
              Object {
                "Group": [Function],
                "IndicatorsContainer": [Function],
                "MenuList": [Function],
                "Option": [Function],
                "SingleValue": [Function],
              }
            }
            isClearable={false}
            isDisabled={false}
            isLoading={false}
            isMulti={false}
            isSearchable={true}
            maxMenuHeight={300}
            onChange={[Function]}
            openMenuOnFocus={false}
            options={Array []}
            tabSelectsValue={true}
          />
        </div>
      </div>
    </div>
    <div
      className="gf-form-inline"
    >
      <div
        className="gf-form"
      >
        <div
          className="max-width-30 gf-form-inline"
        >
          <Button
            disabled={true}
            onClick={[Function]}
            size="sm"
            type="button"
            variant="secondary"
          >
            Load Workspaces
          </Button>
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`Render should enable azure log analytics load workspaces button 1`] = `
<Fragment>
  <h3
    className="page-heading"
  >
    Azure Monitor Logs Details
  </h3>
  <Switch
    checked={false}
    label="Same details as Azure Monitor API"
    onChange={[Function]}
  />
  <AzureCredentialsForm
    credentials={
      Object {
        "azureCloud": "azuremonitor",
        "clientId": "44693801-6ee6-49de-9b2d-9106972f9572",
        "clientSecret": undefined,
        "tenantId": "e7f3f661-a933-4b3f-8176-51c4f982ec48",
      }
    }
    defaultSubscription="e3fe4fde-ad5e-4d60-9974-e2f3562ffdf2"
    getSubscriptions={[MockFunction]}
    onCredentialsChange={[Function]}
    onDefaultSubscriptionChange={[Function]}
  />
  <div
    className="gf-form-group"
  >
    <div
      className="gf-form-inline"
    >
      <div
        className="gf-form"
      >
        <FormLabel
          className="width-12"
          tooltip="Choose the default/preferred Workspace for Azure Log Analytics queries."
        >
          Default Workspace
        </FormLabel>
        <div
          className="width-25"
        >
          <Select
            allowCustomValue={false}
            autoFocus={false}
            backspaceRemovesValue={true}
            className=""
            components={
              Object {
                "Group": [Function],
                "IndicatorsContainer": [Function],
                "MenuList": [Function],
                "Option": [Function],
                "SingleValue": [Function],
              }
            }
            isClearable={false}
            isDisabled={false}
            isLoading={false}
            isMulti={false}
            isSearchable={true}
            maxMenuHeight={300}
            onChange={[Function]}
            openMenuOnFocus={false}
            options={Array []}
            tabSelectsValue={true}
          />
        </div>
      </div>
    </div>
    <div
      className="gf-form-inline"
    >
      <div
        className="gf-form"
      >
        <div
          className="max-width-30 gf-form-inline"
        >
          <Button
            disabled={true}
            onClick={[Function]}
            size="sm"
            type="button"
            variant="secondary"
          >
            Load Workspaces
          </Button>
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`Render should render component 1`] = `
<Fragment>
  <h3
    className="page-heading"
  >
    Azure Monitor Logs Details
  </h3>
  <Switch
    checked={false}
    label="Same details as Azure Monitor API"
    onChange={[Function]}
  />
  <AzureCredentialsForm
    credentials={
      Object {
        "azureCloud": "azuremonitor",
        "clientId": undefined,
        "clientSecret": undefined,
        "tenantId": "",
      }
    }
    getSubscriptions={[MockFunction]}
    onCredentialsChange={[Function]}
    onDefaultSubscriptionChange={[Function]}
  />
  <div
    className="gf-form-group"
  >
    <div
      className="gf-form-inline"
    >
      <div
        className="gf-form"
      >
        <FormLabel
          className="width-12"
          tooltip="Choose the default/preferred Workspace for Azure Log Analytics queries."
        >
          Default Workspace
        </FormLabel>
        <div
          className="width-25"
        >
          <Select
            allowCustomValue={false}
            autoFocus={false}
            backspaceRemovesValue={true}
            className=""
            components={
              Object {
                "Group": [Function],
                "IndicatorsContainer": [Function],
                "MenuList": [Function],
                "Option": [Function],
                "SingleValue": [Function],
              }
            }
            isClearable={false}
            isDisabled={false}
            isLoading={false}
            isMulti={false}
            isSearchable={true}
            maxMenuHeight={300}
            onChange={[Function]}
            openMenuOnFocus={false}
            options={Array []}
            tabSelectsValue={true}
          />
        </div>
      </div>
    </div>
    <div
      className="gf-form-inline"
    >
      <div
        className="gf-form"
      >
        <div
          className="max-width-30 gf-form-inline"
        >
          <Button
            disabled={true}
            onClick={[Function]}
            size="sm"
            type="button"
            variant="secondary"
          >
            Load Workspaces
          </Button>
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;
